packages/modules/notification/src/services/notification-module-service.ts

```diff
@@ -108,57 +109,114 @@ export default class NotificationModuleService
-          throw new MedusaError(
-            MedusaError.Types.NOT_FOUND,
-            `Notification provider ${provider.id} is not enabled. To enable it, configure it as a provider in the notification module options.`
+        const channels = notificationsToProcess.map((not) => not.channel)
+        const providers =
+          await this.notificationProviderService_.getProviderForChannels(
+            channels
           )
-        }
 
-        const res = await this.notificationProviderService_.send(
-          provider,
-          entry
+        // Create the notifications to be sent to prevent concurrent actions listing the same notifications
+        const normalizedNotificationsToProcess = notificationsToProcess.map(
+          (entry) => {
+            const provider = providers.find((provider) =>
+              provider?.channels.includes(entry.channel)
+            )
+
+            return {
+              provider,
+              alreadyExists: !!(
+                entry.idempotency_key && existsMap.has(entry.idempotency_key)
+              ),
+              data: {
+                id: generateEntityId(undefined, "noti"),
+                ...entry,
+                provider_id: provider?.id,
+              },
+            }
+          }
+        )
+
+        const createdNotifications = await this.notificationService_.create(
+          normalizedNotificationsToProcess
+            .filter((e) => !e.alreadyExists)
+            .map((e) => e.data),
+          context
         )
-        return { ...entry, provider_id: provider.id, external_id: res.id }
+
+        return {
+          notificationsToProcess: normalizedNotificationsToProcess,
+          createdNotifications,
+        }
       })
-    )
 
-    // Currently we store notifications after they are sent, which might result in a notification being sent that is not registered in the database.
-    // If necessary, we can switch to a two-step process where we first create the notification, send it, and update it after it being sent.
-    const createdNotifications = await this.notificationService_.create(
-      notificationsToCreate,
-      sharedContext
-    )
+    const notificationToUpdate: { id: string; external_id?: string }[] = []
+
+    try {
+      await promiseAll(
+        notificationsToProcess.map(async (entry) => {
+          const provider = entry.provider
+
+          if (!provider?.is_enabled) {
+            entry.data.status = NotificationStatus.FAILURE
+            notificationToUpdate.push(entry.data)
+
+            const errorMessage = !provider
+              ? `Could not find a notification provider for channel: ${entry.data.channel}`
+              : `Notification provider ${provider.id} is not enabled. To enable it, configure it as a provider in the notification module options.`
+
+            throw new MedusaError(MedusaError.Types.NOT_FOUND, errorMessage)
+          }
+
+          const res = await this.notificationProviderService_.send(
```

Comment:
Should we have a catch here and set the status to failure?

Reply:
We already update all the status in the finally block for both success and failure and we do not have to re throw or anything as it is handled automatically

Reply:
Let me know if you still think that i am missing something. But i ve added tests including both success and failure in order to validate the behavior

Reply:
We do the update there, but we never set the status to `Failure` in case the `.send` fails right? The test only covers the scenario where the provider isn't enabled/doesn't exist. Let me know if I misunderstand the code though

Reply:
Ouch no you are completely right, ill fix that and add a test as well 👌